rosbag2_cpp/src/rosbag2_cpp/writers/sequential_writer.cpp

```diff
@@ -56,16 +58,34 @@ SequentialWriter::SequentialWriter(
   storage_(nullptr),
   metadata_io_(std::move(metadata_io)),
   converter_(nullptr),
+  compressor_{nullptr},
   max_bagfile_size_(rosbag2_storage::storage_interfaces::MAX_BAGFILE_SIZE_NO_SPLIT),
   topics_names_to_info_(),
-  metadata_()
-{}
+  metadata_(),
+  compression_mode_{CompressionMode::NONE} {}
+
 
 SequentialWriter::~SequentialWriter()
 {
   reset();
 }
 
+void SequentialWriter::init_compression(const CompressionOptions & compression_options)
+{
+  if (compression_options.compression_mode != rosbag2_cpp::CompressionMode::NONE) {
+    if (compression_options.compression_format == "zstd") {
+      compressor_ = std::make_unique<rosbag2_compression::ZstdCompressor>();
```

Comment:
I am having a problem with this line as it's imposing a hard dependency on `Zstd` and all it's dependencies. 
`rosbag2_cpp` should potentially be compiled without any compression algorithm

Reply:
We have [this issue](https://github.com/ros-security/aws-roadmap/issues/154) on our backlog next to make compression a plugin instead so that its not hardcoded.

Do you have any other suggestions in the meantime on how to avoid the hard dependency?
Maybe something along the lines of:

```cpp
auto zstd_compressor = std::make_unique<rosbag2_compression::ZstdCompressor>();
compressor_ = std::make_unique<rosbag2_compression::BaseCompressorInterface>(std::move(zstd_compressor));
```

If not, I suggest we work with the current implementation and then after introducing the plugin changes to `rosbag2_compression` use a `CompressorFactory` to make the writer more robust for different compression implementations.

Reply:
Not sure I get your proposition right. We can either use a factory for loading it as a plugin dynamically or as discussed in a previous PR, inherit from the `SequentialWriter` or the `BaseReaderInterface` and make it live in the compression package. It can still be passed in by the caller side when instantiating the `final` Reader class. Example here: https://github.com/ros2/rosbag2/blob/master/rosbag2_transport/src/rosbag2_transport/player.cpp#L44-L46

The moment we include a compression header in the rosbag2_cpp package we introduce the dependency. That being said, the current PR also lacks the dependency of `rosbag2_compression` in its `package.xml`.

Reply:
If I understand correctly, you prefer to isolate compression from `rosbag2_cpp`.
So instead there would be a `CompressionWriter` in `rosbag2_compression` that would implement compression by message or file which can be used in the other `rosbag2` packages.
However, that means losing the ability to split and compress simultaneously since we're overriding `write()` of the `SequentialWriter`.
Moreover, I don't see the issue of introducing compression as a dependency since we'd like to have that feature available for those using both the API and the CLI.

Reply:
Introducing the compression functionality is not the problem. So things like `compression_options` and general compression related structs might very well be living in this package. However, hard coding `ztd` is giving people no change to not compile zstd and its vendor package etc. This might not suit very well to every use case. 
Just the sqlite3 it is possible to compile rosbag2 without any notion of sqlite3.

Reply:
> However, hard coding ztd is giving people no change to not compile zstd and its vendor package etc. 

I agree, which is why we want to work on making compression a plugin and work like `storage` and `storage_factory` does as opposed to the current implementation which is only `Zstd`.

However, since this is the first iteration, we want to make sure the whole compression pipeline works with a single compression format before extending it to other formats though a plugin.

Reply:
> P.S. Accidentally edited your comment, my bad.

That's one way to steer the direction of aa conversation ;-)

> However, since this is the first iteration, we want to make sure the whole compression pipeline works with a single compression format before extending it to other formats though a plugin.

That's fine. I think having a compression writer/reader implementation within `rosbag2_compression` is then the right way to go. It gives you the freedom to test your compression pipeline without introducing a dependency to `zstd` in `rosbag2_cpp`.

Reply:
> I think having a compression writer/reader implementation within `rosbag2_compression` is then the right way to go.

What about splitting then? Basically implementing a `SequentialCompressionWriter`?
If that's the case we can consider going down that route instead for now.